src/meltano/core/project.py

```diff
@@ -563,7 +564,7 @@ def plugin_dir(
         """
         return self.meltano_dir(
             plugin.type,
-            plugin.name,
+            plugin.venv_name if isinstance(plugin, ProjectPlugin) else plugin.name,
```

Comment:
Any better alternatives to this?

Reply:
Hmm, I don't love the fact the directory would depend on the plugin class passed to this method, i.e. whether it's a fully resolved `ProjectPlugin` or a mere `PluginRef`. I searched for references of `Project.plugin_dir` and what types it's passed:

1. `ProjectPlugin`

   https://github.com/meltano/meltano/blob/07764c8f6f51b5174a8b12564a35ba3e34bd6f12/src/meltano/core/plugin_location_remove.py#L188

2. `BasePlugin`, which ultimately magically (via `__getattr__ -> getattr`) refers to `PluginDefinition`, another subclass of `PluginRef`

   https://github.com/meltano/meltano/blob/07764c8f6f51b5174a8b12564a35ba3e34bd6f12/src/meltano/core/plugin/file.py#L69

3. `ProjectPlugin`

   https://github.com/meltano/meltano/blob/07764c8f6f51b5174a8b12564a35ba3e34bd6f12/src/meltano/core/plugin_invoker.py#L171

4. `ProjectPlugin`

   https://github.com/meltano/meltano/blob/07764c8f6f51b5174a8b12564a35ba3e34bd6f12/src/meltano/core/plugin_install_service.py#L419

Reply:
We could give `PluginRef` a naive implementation of `venv_name` such we don't need to do a type check here:

```python
@property
def venv_name(self):
    self.name
```

That still doesn't solve

> the directory would depend on the plugin class passed to this method

though 🤔

Reply:
> 2. `BasePlugin`, which ultimately magically (via `__getattr__ -> getattr`) refers to `PluginDefinition`, another subclass of `PluginRef`
>    https://github.com/meltano/meltano/blob/07764c8f6f51b5174a8b12564a35ba3e34bd6f12/src/meltano/core/plugin/file.py#L69

Can we resolve this to a `ProjectPlugin` and type hint `ProjectPlugin` in the `plugin_dir` signature instead?

Reply:
Yeah, I'd favor that if we can avoid circular imports

Reply:
2418b42

Let's see how the tests do...

Reply:
99e2314 worked in the end. I guess a file plugin might not necessarily exist in the project in the project at the point `plugin_dir` is called there?

Reply:
Yeah, some file plugins are not permanently added to the project after they're installed. They're only added if they can _update_ their files 🤷‍♂️. I'll be happy if we rip them off the code base 😅.